docs/docs/adrs/adr-018-permissionless-ics.md

```diff
@@ -0,0 +1,270 @@
+| `MinimumPowerInTopNBytePrefix`           |Stores the minimum power needed to opt in for a chain                             |**YES**                   |Every epoch                                                               |
+| `ConsumerAddrsToPruneV2BytePrefix`       |Stores consumer addresses to be pruned (as part of `VSCMaturedPacket`s deprecation)|**YES**                   |Every `MsgAssignConsumerKey` or `MsgOptIn` and later during actual pruning|
+
+Everything stored under a key associated with a `chainID` needs to be migrated to new state under `consumerID`. 
+Because we have to migrate in any case, we can also clean up a number of those keys by building a `ConsumerChainRecord`.
+The `ConsumerChainRecord` contains state relevant to a consumer chain and is keyed by a `consumerID`. Although the `ConsumerChainRecord`
+could contain **all** state related to a consumer chain (e.g., opted-in and consumer validators of a chain) we do not include
+such fields in the `ConsumerChainRecord` because this would increase the cost of ICS-related transactions due to the [gas cost](https://github.com/cosmos/cosmos-sdk/blob/v0.50.7/store/gaskv/store.go#L40).
+Furthermore, if we were to store all the opted-in or consumer validators, etc. it would be tricky to read or write those fields,
+because we would need to extract the opted-in validators from the `ConsumerChainRecord` and then search through those
+validators to find the one we are looking for, etc.
+
+The `ConsumerChainRecord` contains the following:
+```protobuf
+message ConsumerChainRecord {
+  // the owner of this consumer chain
+  string owner_address;
+  // client id to the consumer chain
+  string client_id;
+  // channel id
+  string channel_id;
+  // the chain-id of the consumer chain
+  string chain_id;
+  // consumer genesis
+  interchain_security.ccv.v1.ConsumerGenesisState consumer_genesis = 5
+  // min height needed for an equivocation to be actionable
+  uint64 equivocation_evidence_min_height;
+  // maximum power (percentage-wise) a validator can have on the consumer chain
+  uint32 validators_power_cap;
+  // maximum number of validators that can validate a consumer chain
+  uint32 validator_set_cap;
+  // allowlist of provider consensus addresses of validators that are the ONLY ones that can validate the consumer chain
+  repeated string allowlist;
+  // denylist of provider consensus addresses of validators that are the CANNOT validate the consumer chain
+  repeated string denylist;
+}
+```
+
+We store the `ConsumerChainRecord`s using the `consumerID` as their key. The `owner_address` of a consumer chain corresponds
+to the address that issued the `MsgLaunchConsumerChain` and the `owner_address` can be updated by issuing a `MsgUpdateConsumerChain` (see later).
+Note that a `ConsumerChainRecord` is only created when a consumer chain moves from the prelaunched to the _launched_ phase.
```

Comment:
what is the reason to store it in advance as a proposal?
are there any side-effects doing so, i.e. query on gov proposals will show up these entries as well?

Reply:
The reason is to simplify implementation. `ConsumerAdditionProposal` already exists and contains everything we need. As a result, we would have less code changes.
> are there any side-effects doing so, i.e. query on gov proposals will show up these entries as well?
I do not see this being the case because we manually create `ConsumerAdditionProposal` and hall `HandleConsumerAdditionProposal`. `HandleConsumerAdditionProposal` is in gnerally called after a proposal has already passed. In any case, I'll take another look on this.

Reply:
Why would this complicate things? Note that `ConsumerAdditionProposal` contains substantial more data (e.g., `spawnTime`) than what the `ConsumerChainRecord` ends up having.

I think it would make sense to rename `ConsumerAdditionProposal` to `PrelaunchConsumerChainRecord` and reuse it for gov proposals and the `MsgLaunchConsumerChain` message that would probably need to be renamed to `MsgPrelaunchConsumerChain`.